### ISSUE TITLE
Low success metrics under load

### ISSUE THREAD

**Issue by agniveshadhikari** [linkerd/linkerd2#7768]
### What is the issue?

I am using `hey` to generate some load on my system, with the options `-z 60s -c 100`. Summary given by hey:
```
Status code distribution:
    [200]       424 responses
    [500]       727 responses
    [503]       289 responses
```
My "system" is one pod of ingress-nginx making calls to this service X(=data-plane--dataplane-service) that I am load testing. X has only one pod and no HPA.

These are some graphs from linkerd viz for the deploy/X:
![image](https://user-images.githubusercontent.com/23560829/152222887-2407a239-5485-4ef2-badd-a4af9dca7b26.png)

I can see WARN logs from the linkerd-proxy container on the pod for service X saying HTTP Balancer service and HTTP Logical service are entering failfast, and then after some time, in a loop. I've included some logs below.

I don't see any logs from the proxy container of nginx pod.

I'm happy to provide any other details that might help.

### How can it be reproduced?

Using a command something like this to generate the load:
`hey -z 60s -c 100 ingress-nginx.myservice.com`

Here is the part of my cluster that we are talking about:
![image](https://user-images.githubusercontent.com/23560829/152225782-076f59e0-e925-482f-856b-926c55074745.png)

Annotations on the ingress-nginx pod:
```
  Annotations:      config.linkerd.io/skip-inbound-ports: 80,443
                    linkerd.io/inject: enabled
```

Annotations on the target service X:
```
  Annotations:  config.linkerd.io/skip-outbound-ports: 2579
                linkerd.io/inject: enabled
```

### Logs, error output, etc

```
[560294.049116s]  INFO ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_proxy_api_resolve::resolve: No endpoints
[560297.148457s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_timeout::failfast: HTTP Logical entering failfast after 3s
[560297.149165s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_timeout::failfast: HTTP Balancer entering failfast after 3s
[560297.149285s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Logical service in fail-fast client.addr=10.244.7.178:59510
[560297.149378s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Logical service in fail-fast client.addr=10.244.7.178:59512
[560297.149456s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Logical service in fail-fast client.addr=10.244.7.178:59514
# multiple lines omitted
[560297.154943s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Logical service in fail-fast client.addr=10.244.7.178:59254
[560297.154981s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Logical service in fail-fast client.addr=10.244.7.178:59202
[560297.155045s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Logical service in fail-fast client.addr=10.244.7.178:59392
[560298.228210s]  INFO ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_timeout::failfast: HTTP Logical service has recovered
[560298.228278s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Balancer service in fail-fast client.addr=10.244.7.178:59142
[560298.228612s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Balancer service in fail-fast client.addr=10.244.7.178:59420
[560298.228866s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Balancer service in fail-fast client.addr=10.244.7.178:59144
# multiple lines omitted
[560299.037395s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Balancer service in fail-fast client.addr=10.244.7.178:60028
[560299.041354s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Balancer service in fail-fast client.addr=10.244.7.178:60030
[560299.041710s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Balancer service in fail-fast client.addr=10.244.7.178:60032
[560299.167755s]  INFO ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_timeout::failfast: HTTP Balancer service has recovered
[560312.064706s]  INFO ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_proxy_api_resolve::resolve: No endpoints
[560315.224184s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_timeout::failfast: HTTP Balancer entering failfast after 3s
[560315.225130s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Balancer service in fail-fast client.addr=10.244.7.178:60206
[560315.225327s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Balancer service in fail-fast client.addr=10.244.7.178:60160
[560315.225407s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Balancer service in fail-fast client.addr=10.244.7.178:60056
# multiple lines omitted
[560317.032979s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Balancer service in fail-fast client.addr=10.244.7.178:60934
[560317.033045s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Balancer service in fail-fast client.addr=10.244.7.178:60936
[560317.038445s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_app_core::errors: Failed to proxy request: HTTP Balancer service in fail-fast client.addr=10.244.7.178:60940
[560317.120544s]  INFO ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_timeout::failfast: HTTP Balancer service has recovered
[560322.055615s]  INFO ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_proxy_api_resolve::resolve: No endpoints
[560325.256130s]  WARN ThreadId(01) server{orig_dst=10.0.46.101:80}: linkerd_timeout::failfast: HTTP Balancer entering failfast after 3s
# ... and so on
```

### output of `linkerd check -o short`

```
$ linkerd check -o short
Linkerd core checks
===================

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 2.10.2 but the latest stable version is 2.11.1
    see https://linkerd.io/2.11/checks/#l5d-version-control for hints
‼ control plane and cli versions match
    control plane running stable-2.10.2 but cli running stable-2.11.1
    see https://linkerd.io/2.11/checks/#l5d-version-control for hints

linkerd-control-plane-proxy
---------------------------
‼ control plane proxies are up-to-date
    some proxies are not running the current version:
        * linkerd-controller-8b6bdd866-94hg2 (stable-2.10.2)
        * linkerd-controller-8b6bdd866-ls47c (stable-2.10.2)
        * linkerd-controller-8b6bdd866-zj5jb (stable-2.10.2)
        * linkerd-destination-74f855b7dd-5cwg8 (stable-2.10.2)
        * linkerd-destination-74f855b7dd-d5kzh (stable-2.10.2)
        * linkerd-destination-74f855b7dd-gp5wx (stable-2.10.2)
        * linkerd-identity-5f8bf6f6cf-cgb2j (stable-2.10.2)
        * linkerd-identity-5f8bf6f6cf-hwlhr (stable-2.10.2)
        * linkerd-identity-5f8bf6f6cf-n89v8 (stable-2.10.2)
        * linkerd-proxy-injector-76fb75897c-4x86v (stable-2.10.2)
        * linkerd-proxy-injector-76fb75897c-nhrrg (stable-2.10.2)
        * linkerd-proxy-injector-76fb75897c-wwqwn (stable-2.10.2)
        * linkerd-sp-validator-55c899d895-5d7bx (stable-2.10.2)
        * linkerd-sp-validator-55c899d895-gzn5z (stable-2.10.2)
        * linkerd-sp-validator-55c899d895-pql6x (stable-2.10.2)
    see https://linkerd.io/2.11/checks/#l5d-cp-proxy-version for hints
‼ control plane proxies and cli versions match
    linkerd-controller-8b6bdd866-94hg2 running stable-2.10.2 but cli running stable-2.11.1
    see https://linkerd.io/2.11/checks/#l5d-cp-proxy-cli-version for hints

linkerd-ha-checks
-----------------
‼ pod injection disabled on kube-system
    kube-system namespace needs to have the label config.linkerd.io/admission-webhooks: disabled if injector webhook failure policy is Fail
    see https://linkerd.io/2.11/checks/#l5d-injection-disabled for hints

Status check results are √

Linkerd extensions checks
=========================

linkerd-viz
-----------
‼ viz extension proxies are up-to-date
    some proxies are not running the current version:
        * grafana-7c4bc49848-b9wql (stable-2.10.2)
        * metrics-api-558f7995cb-ngmg8 (stable-2.10.2)
        * prometheus-fb9d58f4-skw5l (stable-2.10.2)
        * tap-59c6f9b6f7-fprtw (stable-2.10.2)
        * tap-injector-576b974499-ccxzc (stable-2.10.2)
        * web-79984fd46c-nwbcx (stable-2.10.2)
    see https://linkerd.io/2.11/checks/#l5d-viz-proxy-cp-version for hints
‼ viz extension proxies and cli versions match
    grafana-7c4bc49848-b9wql running stable-2.10.2 but cli running stable-2.11.1
    see https://linkerd.io/2.11/checks/#l5d-viz-proxy-cli-version for hints

Status check results are √
```

### Environment

- Kubernetes version: v1.21.7
- Env: Azure Kubernetes Service
- OS: Linux

### Would you like to work on fixing this bug?

maybe

**Comment by agniveshadhikari** [comment]
Like I said, I'm happy to give you more details. Didn't want to bloat the initial report with too many possibly useless details.

**Comment by wmorgan** [comment]
Given the 10s max latency I suspect you're running into a protocol detection timeout. Is dataplane-service calling an off-cluster database, for example?

**Comment by agniveshadhikari** [comment]
@wmorgan Not directly. But both the other services you see it calling into talk to off cluster DB. I'll have to educate myself on "protocol detection timeout". Is there something data point or test that will help us pinpoint the issue?

**Comment by wmorgan** [comment]
There's a flow chart in https://linkerd.io/2.11/features/protocol-detection/ that you can follow.

If it's happening you'll see a message about it in the logs. (We'll capture it in metrics in 2.12 as well, but that's not there yet. See #6540.)

**Comment by ayushiaks** [comment]
Hi, we don't see any logs related to protocol detection timeout happening here. We also tried to set the ports as opaque to get past it, but the issue still persists.

We regularly see the proxy call failing, as it couldn't reach it's destination:
`[246688.447807s]  INFO ThreadId(01) linkerd_app_core::serve: Connection closed error=connection closed before message completed
[246688.470149s]  WARN ThreadId(01) server{orig_dst=ip-x}: linkerd_app_core::errors: Failed to proxy request: operation was canceled: connection closed client.addr=ip-y`

The call is being made from one of our internal services to the other. And the call is being terminated as
`504 server error: Gateway timeout for url`

The services were in a healthy state meanwhile.
Can we maybe increase the timeout for linkerd before it closes the connection? As the response time might be delayed under load.

**Comment by ayushiaks** [comment]
Hi @wmorgan 
You're correct, looks like we're hitting the protocol detection timeout, while trying to talk to a off cluster database.

```
INFO ThreadId(01) server{orig_dst=<>}: linkerd_detect: Continuing after timeout: linkerd_proxy_http::version::Version protocol detection timed out after 10s
```

Here the destination's IP address rightly points to the database we're trying to call into.
Some ongoing calls to the database also terminated due to timeout.

I tried making the port 80 opaque, through which our service talks, but I'm seeing this error now:

```
linkerd_app_core::serve: Connection closed error=a named target must be provided on gateway connections
```

Any clue how to fix this?

Also, why does it timeout only for off cluster calls? and why under load?

**Comment by kleimkuhler** [comment]
@ayushiaks Marking a port as opaque means that Linkerd will skip protocol detection and proxy the connection as a TCP stream, but it will still encrypt that traffic similar to when it knows the protocol it is working with.

In order to fix your issue, you'll have to mark the port as a skip-port. This is because the destination is off-cluster, so we don't want Linkerd to encrypt the traffic—or do anything with it! This does mean you'll lack TCP metrics for this connection, but it's a tradeoff you'll have to make for off-cluster destinations.

The annotation you'll be looking for is `config.linkerd.io/skip-outbound-ports` and you can apply this to the workload that is talking to the off-cluster database.

You can also read more about the difference between opaque and skip-ports in the documentation for [configuring protocol detection](https://linkerd.io/2.10/features/protocol-detection/#configuring-protocol-detection).

**Comment by ayushiaks** [comment]
@kleimkuhler the protocol detection error does go away with skipping the port (making it opaque causes more errors than solving it, like it should've according to the documentation), but it's still unclear on why this was happening for https calls. Any clues?

**Comment by ayushiaks** [comment]
Well, I think I get it now :) 
The doc mentions that opaque ports can only be used for destinations inside the cluster, and that linkerd will try to detect what the protocol is for off cluster calls as well.

Skipping the port works for me. Thanks! Closing this issue

**Comment by wmorgan** [comment]
@ayushiaks I'm glad you got it working but I'm a little confused. If the calls to the database are already TLS'd by the client, then skip-ports should not be necessary. Linkerd should immediately detect the TLS connection and proxy it through uninterrupted.

If you can create a little repro case where you can made TLS calls that trigger protocol detection timeouts that we can try on our side, I would consider filing a bug report.

**Comment by ayushiaks** [comment]
@wmorgan isn't protocol detection being used for differentiating between https and http calls too?
The documentation is a bit confusing in that sense. It says https calls are read as plain tcp as they can't be decrypted to get http level transactions.

Doesn't linkerd need to detect it's a https protocol first, to decide not to do anything with it?
If not, how does it work? What protocols is it trying to differentiate between?

And yes, let me try to create an easy repro for this meanwhile

**Comment by mateiidavid** [comment]
Hey @ayushiaks, let's see if I can simply everything a little bit.

**Opaque ports**: I see you've sort of understood what the behaviour is like when we use opaque ports. I wanted to add a bit of how the implementation works, maybe it'll help you in the future. Opaque ports are set on a _service_ (logical endpoint), and on a _pod_ (concrete endpoint). 

 * When set on a service, linkerd-proxy will not do any protocol detection when building a logical (load balanced) connection.
 * When set on a pod (endpoint), linkerd-proxy will change the target port. For example, if the request is supposed to go to `8080` but the pod is opaque, the target port will be changed to the inbound proxy's listen port ( `4143`), from there, the inbound proxy will open a TCP socket with your container and forward everything it received from your client.

**Protocol detection**: will happen both on the inbound, and the outbound side. We're primarily concerned with the outbound side here, that's where protocol detection timeouts happen. It's important to note that **a 10 second timeout happens, when the protocol cannot be detected**. Under normal circumstances, it takes a small fraction of time to actually detect the protocol, so in 90% of the cases, there isn't really a timeout to be concerned with.

To understand protocol detection, we can look at the [code](https://github.com/linkerd/linkerd2-proxy/blob/main/linkerd/detect/src/lib.rs#L149-L175). Fortunately, this isn't really complicated, even if you're not knowledgeable with Rust.

* We create a timer
* We create a buffer
* We attempt to detect the protocol before the timer runs out. To perform detection, we pass in the client side of the socket and our buffer. If we successfully detect the protocol, then that's kind of it, otherwise we treat the connection as opaque (if the timer has run out).

**How do we actually detect the protocol?**: we peek at the first few bytes in the client socket (our `IO`). HTTP and gRPC protocols in their spec will specify what the first few lines are, this is mostly the protocol version, path accessed and so on. If the first few bytes indicate the protocol, then that's the protocol detected. With TLS, we'll get a bunch of garbage bits basically, since the payload is encrypted, the connection will be opaque (since the protocol is NOT http 1 or 2) but the timeout will not be hit. On the inbound side, when we detect the protocol, we read the SNI to see whether this has already been TLS'd.

That's how we infer the connection has already been TLS'd. If you're curious, here are some implementations of our detect interface:

[[1](https://github.com/linkerd/linkerd2-proxy/blob/f5e9cea4698cd4cf74f791b74168cda161f70552/linkerd/tls/src/server.rs#L174)]: TLS DETECT
[[2](https://github.com/linkerd/linkerd2-proxy/blob/1cff3aef82c203bf09ccce485506d7a29ca27308/linkerd/proxy/http/src/detect.rs#L31)]: HTTP DETECT 

With other protocols, such as server speak first protocols, we will hit the timeout because **the client socket will not have any data in it**. In those cases, the server will initiate all of the protocol configuration details. We'll try to read off the socket for 10s but nothing will ever be present, so we timeout and go down the opaque route.

---

**TL;DR**

>  isn't protocol detection being used for differentiating between https and http calls too?
The documentation is a bit confusing in that sense. It says HTTPS calls are read as plain TCP as they can't be decrypted to get HTTP level transactions.

Yes, protocol detection is used, but the result is instant, there's no timeout.

> Doesn't linkerd need to detect it's a HTTPS protocol first, to decide not to do anything with it?
If not, how does it work? What protocols is it trying to differentiate between?

On the outbound side we sort of establish that it's not HTTP. On the inbound side, we read the SNI. 

This is my understanding of what's going on, ofc open to being corrected by others :)

**Comment by ayushiaks** [comment]
Thanks @mateiidavid for the detailed explanation. It does make sense, but 

`With TLS, the connection will be opaque but the timeout will not be hit`

The connection will be opaque -  also seems like a part of protocol detection as you explained.
How are we sure the timeout won't be hit.

Since I'm clearly able to see that my TLS off cluster calls are the only ones hitting detection timeout, I'm worried that might not be true.
Ofcourse for 90% calls its instant, this behaviour is only observed under load.

Sample of the logline where I see the error:
`server{orig_dst=ip:443}: linkerd_detect: Continuing after timeout: linkerd_proxy_http::version::Version protocol detection timed out after 10s`

We know this endpoint is TLS'd, and is using the conventional 443 port for https calls.

**Comment by ayushiaks** [comment]
Hi @mateiidavid @wmorgan 
I cannot repro the issue anymore, has anything changed?

**Comment by mateiidavid** [comment]
@ayushiaks sorry for a very tardy response on my side. Chances are nothing has been changed, especially if you run the same linkerd version. We haven't added (or modified) the proxy's behaviour around protocol detection in the past few edge versions though.

> The connection will be opaque - also seems like a part of protocol detection as you explained.
How are we sure the timeout won't be hit while reading the garbage bits?

Because data will always be present in the client socket when doing TLS. Whenever TLS is used, the client first sends a few bytes representing a `ClientHello` -- this is used in the TLS handshake. By contrast, in a Server Speaks First protocol, we won't have any bytes in the **client socket** since we expect the server to send the protocol set-up bytes.

Are you still seeing this issue though? Would be more than happy to help you troubleshoot further.
